main.go

```diff
@@ -14,191 +14,336 @@ import (
-	client := http.Client{Timeout: 3 * time.Second}
-
-	resp, err := client.Get(e.Target + "/monitor/statistics.json")
-	if err != nil {
-		return err
+func NewMesosExporter(opts *ExporterOpts) *PeriodicExporter {
+	e := &PeriodicExporter{
+		errors: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "exporter",
+				Name:      "scrape_errors",
+				Help:      "Current total scrape errors ",
+			},
+			[]string{"mesos_slave"},
+		),
+		opts:    opts,
+		metrics: make([]prometheus.Gauge, 0),
 	}
+	e.slaves.hostnames = []string{e.opts.localAddr}
 
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return err
+	if e.opts.autoDiscover {
+		glog.Info("auto discovery enabled from command line flag.")
+
+		// Update nr. of mesos slave every 10 minute
+		e.slaves.hostnames = []string{}
+		go e.runEvery(e.updateSlaves, (10 * time.Minute))
 	}
 
-	var stats []mesos_stats.Monitor
+	// Fetch slave metrics every interval
+	go e.runEvery(e.scrapeSlaves, e.opts.interval)
 
-	err = json.Unmarshal(body, &stats)
-	if err != nil {
-		return err
-	}
+	return e
+}
 
-	for _, t := range stats {
-		cpuLimitVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusLimit)
-		cpuSysVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusSystemTimeSecs)
-		cpuUsrVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusUserTimeSecs)
-		memLimitVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(float64(t.Statistics.MemLimitBytes))
-		memRssVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(float64(t.Statistics.MemRssBytes))
-	}
+func (e *PeriodicExporter) Describe(ch chan<- *prometheus.Desc) {
+	e.rLockMetrics(func() {
+		for _, m := range e.metrics {
+			m.Describe(ch)
+		}
+	})
+	e.errors.MetricVec.Describe(ch)
+}
 
-	return err
+func (e *PeriodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.rLockMetrics(func() {
+		for _, m := range e.metrics {
+			m.Collect(ch)
+		}
+	})
+	e.errors.MetricVec.Collect(ch)
 }
 
-func requestRunner(e Endpoint) {
-	for {
-		err := request(e)
+func (e *PeriodicExporter) fetch(inCh chan string, outCh chan prometheus.Gauge, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for host := range inCh {
+		monitorURL := fmt.Sprintf("http://%s:5051/monitor/statistics.json", host)
+		resp, err := httpClient.Get(monitorURL)
 		if err != nil {
-			glog.Warningln(err)
+			glog.Warningf("GET %s failed. Error: %s", monitorURL, err)
+			e.errors.WithLabelValues(host).Inc()
+			continue
 		}
+		defer resp.Body.Close()
 
-		t, err := time.ParseDuration(e.Interval)
-		if err != nil {
-			glog.Warningln(err)
-			t = 30 * time.Second
+		var stats []mesos_stats.Monitor
+		if err := json.NewDecoder(resp.Body).Decode(&stats); err != nil {
+			glog.Warningf("failed to deserialize request: %s", err)
+			e.errors.WithLabelValues(host).Inc()
+			continue
 		}
 
-		time.Sleep(t)
+		for _, stat := range stats {
+			cpuLimitVal := prometheus.NewGauge(
```

Comment:
Since all these metrics are instantiated "throw-away" style (i.e. not kept around for later runs), it would be more common to use `ConstMetrics` in this use case. An example of that:

https://github.com/prometheus/client_golang/blob/de5f7a2db9d883392ce3ad667087280fe1ff9cea/prometheus/example_memstats_test.go#L59-L80